### PR TITLE
[FIX] hr_expense: Add inverse on total_amount to update unit_amount

### PR DESCRIPTION
**Pull request by ryce-odoo**
[FIX] hr_expense: Add inverse on total_amount to update unit_amount
The customer was having an issue where they would create an expense from an email with the wrong cost on the expense.
Then when they would try to change the price on the expense record it would seem find all the way up until the journal was posted.

Issue:
When creating an expense from an email the unit_amount is updated instead of the total_amount.
This would cause issues if you had an attachment because when you would attempt to change the amount through the front-end this unit_amount would never get updated because it was set and there was an attachment (line 265).
However, when you would create a journal entry from this expense, because it has the unit_amount != 0 it would provide the unit_amount instead of the total_amount.
Thus propogating the original number from the email even if it was updated between the time of posting and the creation of the expense from the email.

Solution:
Implement an inverse function on total_amount that will update the unit_amount to total_amount_company.
This will cause the unit_amount to stay up to date with the total_amount while also taking into account currency.

opw-3286372
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ryce-odoo** [comment]
@jbw-odoo Can you please check this? Customer is awaiting a resolution so that they can send expenses through the mail alias again.

**Comment by jbw-odoo** [comment]
@Levizar

**Comment by Levizar** [comment]
Hello @ryce-odoo , sorry for that LONG STUFF

### TLDR: 
- I think the fix isn't ok because it breaks a feature
- We should probably use an inverse method for `total_amount` with `@onchange` on it
- Would be nice to add a test

### BIG BLABLA:
I may be wrong but I think the solution isn't ok.
Changing that part means the line 265 that you point out becomes pointless.

I am nearly 100% sure that the goal of the initial code was to not modify the `unit_amount` by changing the product if it was provided by the attachment.
With the current fix, that won't happen anymore which could be very annoying for a user that would just want to change the product.

To me, the real issue is that the customer can't change it even if they change the `total_amount`.

Thus, I think a better solution would be to put an `inverse` method for the `total_amount` field that would recompute the `unit_amount`.
That would make the `unit_amount` change if the user modifies the `total_amount`.
if we need that to be dynamically modified in the UI, we could add an `onchange` on that inverse method. (because inverse are only triggered when the model is saved)

Also, very important part: could you add a test?

**Comment by ryce-odoo** [comment]
Hey @Levizar,

This does sound like it could be a better solution. However, I am a bit confused as to what purpose the unit_amount serves exactly. As why would you not want to change it if the product is provided by the email? This seems like a possible oversight as there is code that will provide the total_amount instead of the unit_amount to the journal entry if the unit_amount is 0. 

From what I could tell, I believe that the unit_amount should stay 0 instead of writing to it if the product does not have a unit_amount otherwise that line checking if unit_amount is 0 serves no purpose.

For a full solution, I believe that both of these changes should be implemented, but the inverse should only write to the unit_amount if it is not already 0. This seems to keep the line of code involving checking unit_amount usable, but I am not too sure on what purpose unit_amount serves in the first place if it will always reflect total_amount.

**Comment by jbw-odoo** [comment]
> However, I am a bit confused as to what purpose the unit_amount serves exactly. As why would you not want to change it if the product is provided by the email? 

If you assume it was correctly extracted form an email, it must not be replaced by the price from a product that might be different and incorrect ?
Maybe implement the inverse as a temp fix and bring this to the attention of a PO ?
@Levizar ?

**Comment by ryce-odoo** [comment]
@jbw-odoo @Levizar 

Overall, it seems like this model needs some refactoring. However, a quick band-aid fix is implementing the inverse. In this case however, the tax must be accounted for, so the in the event that the total_amount is changed it will properly populate the unit_amount to give when preparing move lines.

The solution I am thinking of in this case is to iterate through all tax_ids, then calculate the taxes in reverse to find the unit_amount. To account for tax groups, BFS to find all child tax_ids and use those in the calculation.

unit = total/(1+sum(taxes))

What are your thoughts on this solution?

**Comment by Levizar** [comment]
@ryce-odoo I would try without trying anything regarding tax first to check what are the results on the generated accounting entry.
The reason is that the tax used on such expense are supposed to be VAT included.
If we provide a total amount VAT included and the tax used is VAT included, the accounting entry should take that into account on its own.

**Comment by ryce-odoo** [comment]
Hey @Levizar ,

I have implemented the inverse function on total_amount. This seems to be the best way for now to get the desired functionality when attachments are in play. However, this model may need some refactoring to come up with a long term solution.

If you could check out my most recent changes, that would be much appreciated. 
I have also included a test as requested.

**Comment by jbw-odoo** [comment]
@JulienAlardot

File: addons/hr_expense/models/hr_expense.py
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:323]:
```suggestion
            expense.unit_amount = expense.total_amount_company / expense.quantity
```
As discussed, debated and thought deeply (including the PO)
 - The fix as it is fine (would  make the `unit_amount = total_amout / quantity` in the inverse) even though it shouldn't change much anyway
 - The refactor of the module is getting increasingly more urgent, thank you for getting me more arguments in that fight :p

Comment (ryce-odoo) [reply addons/hr_expense/models/hr_expense.py]: Thanks @JulienAlardot!
I have committed the suggested changes.

**Comment by Levizar** [comment]
runbot is red
odoo.addons.hr_expense.tests.test_expenses:744
                                    ERROR: TestExpenses.test_hr_expense_split
Traceback (most recent call last):
  File "/data/build/odoo/addons/hr_expense/tests/test_expenses.py", line 744, in test_hr_expense_split
    split_wizard = expense.action_split_wizard()
  File "/data/build/odoo/addons/hr_expense/models/hr_expense.py", line 551, in action_split_wizard
    splits = self.env['hr.expense.split'].create(self._get_split_values())
  File "<decorator-gen-329>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 410, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/addons/analytic/models/analytic_mixin.py", line 83, in create
    return super().create(vals_list)
  File "<decorator-gen-65>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 410, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/odoo/addons/base/models/ir_fields.py", line 665, in create
    recs = super().create(vals_list)
  File "<decorator-gen-13>", line 2, in create
  File "/data/build/odoo/odoo/api.py", line 410, in _model_create_multi
    return create(self, arg)
  File "/data/build/odoo/odoo/models.py", line 3902, in create
    raise ValueError("Invalid field %r on model %r" % (key, self._name))
ValueError: Invalid field 'unit_amount' on model 'hr.expense.split'

**Comment by ryce-odoo** [comment]
@Levizar 
I have fixed the issue. Let me know if there are any other problems.

**Comment by Levizar** [comment]
@ryce-odoo It seems that you are cursed :rofl: 

[odoo/addons/hr_expense/tests/test_expenses.py:954](https://github.com/odoo/odoo/blob/23f82d96637de8d9eb4f8dd93f4c06c4e94938b6/addons/hr_expense/tests/test_expenses.py#L954)
```
self.assertEqual(expense.unit_amount, 90, 'Unit amount should be the same as total amount was written to')
C0304(missing-final-newline) coll 0: Final newline missing
```

File: addons/hr_expense/models/hr_expense.py
Comment (Levizar) [inline addons/hr_expense/models/hr_expense.py:324]: @JulienAlardot Didn't we say we need an `@api.onchange('total_amount')` here ?

Also, I'm wondering why we use `total_amount_company` instead of `total_amount`.
Doesn't the `hr_expense` take care of the currencies?

File: addons/hr_expense/tests/test_expenses.py
Comment (Levizar) [inline addons/hr_expense/tests/test_expenses.py:954]: The CI style is failing because of a missing EOF
```suggestion
        self.assertEqual(expense.unit_amount, 90, 'Unit amount should be the same as total amount was written to')

```

**Review by Levizar — COMMENTED** [review]
I'm waiting for JUAL's answer before taking the final decision

File: addons/hr_expense/models/hr_expense.py
Comment (JulienAlardot) [reply addons/hr_expense/models/hr_expense.py:324]: True, we need that onchange
AS it is, no. Currencies for expense are a shitstorm of weird code iteratively changed back and forth. But I'm working on that in another fix

**Comment by JulienAlardot** [comment]
> @ryce-odoo It seems that you are cursed rofl
> 
> [odoo/addons/hr_expense/tests/test_expenses.py:954](https://github.com/odoo/odoo/blob/23f82d96637de8d9eb4f8dd93f4c06c4e94938b6/addons/hr_expense/tests/test_expenses.py#L954)
> 
> ```
> self.assertEqual(expense.unit_amount, 90, 'Unit amount should be the same as total amount was written to')
> C0304(missing-final-newline) coll 0: Final newline missing
> ```

It fails because product_a has a unit_amount, then overrides everything. Use product_c

File: addons/hr_expense/models/hr_expense.py
Comment (ryce-odoo) [reply addons/hr_expense/models/hr_expense.py:325]: Sounds good I have implemented the onchange

File: addons/hr_expense/tests/test_expenses.py
Comment (ryce-odoo) [reply addons/hr_expense/tests/test_expenses.py]: Added EOF newline

**Comment by ryce-odoo** [comment]
> Also, I just noticed: did you see that there is both a `unit_amount` and a `unit_amount_display` ?

I did see this, but I saw that the unit_amount_display compute has some conditions as to what to set it as, so I did not want to forcefully write over it.

Thanks for all the help guys.

Let me know if I get cursed again 😅

@Levizar @JulienAlardot

**Comment by Levizar** [comment]
Hello @ryce-odoo you need to get rid of the merge commit and to quash the 2 others.

**Comment by ryce-odoo** [comment]
Hey @Levizar

It should be good to go now!

**Comment by ryce-odoo** [comment]
@Levizar 
Conflict is now resolved after some github merge shenanigans. Not gonna use github merge again.......💀💀💀

**Comment by Levizar** [comment]
@ryce-odoo and now it's the CI style 🤣

**Comment by ryce-odoo** [comment]
@Levizar 
It's so cursed :(
This should be it now?
Unless there's more merge conflicts....

**Comment by Levizar** [comment]
@ryce-odoo I think the commit message isn't accurate anymore.

@JulienAlardot @jbw-odoo has one of you test the flow?

**Comment by ryce-odoo** [comment]
Hello @Levizar,

All good I should have updated it when squashing. I have now updated the commit message, title, and my initial comment now as to not cause any confusion.

Let me know if I need to do anything else!

**Review by Levizar — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**
